### PR TITLE
#145 import image series

### PR DESCRIPTION
**Pull request by vessemer**
Import image series implementation. 

## Reference to official issue
This addresses #145 

## Screenshots:
ezgif com-cro

## CLA
- [x] I have signed the CLA; if other committers are in the commit history, they have signed the CLA as well

### CODE REVIEW

File: interface/frontend/src/components/open-image/ImageSeries.vue [hunk 1]
```diff
@@ -80,6 +87,7 @@
   import { EventBus } from '../../main.js'
   import TreeView from './TreeView'
   import OpenDicom from './OpenDICOM'
+  const dirname = require('path-dirname')
```
Comment (louisgv) [inline interface/frontend/src/components/open-image/ImageSeries.vue]: Hmm, seems like this module cannot be used with `import`? I would imagine `import dirname from 'path-dirname'` woudl work.

Comment (vessemer) [reply interface/frontend/src/components/open-image/ImageSeries.vue]: Well, yes, it works like you thought. I was confused by this [this](https://github.com/concept-to-clinic/concept-to-clinic/pull/188#discussion_r147443976) obstacle for the `cornerstone` and `cornerstone-tools`.

**Comment by Serhiy-Shekhovtsov** [comment]
@vessemer, I think it's a good idea to let others know when you are working on an issue (like I did [here](https://github.com/concept-to-clinic/concept-to-clinic/issues/145#issuecomment-344731667)) if we are choosing collaboration over competition. This way team members will not waste their time working on same tasks.

**Comment by Serhiy-Shekhovtsov** [comment]
Image loading is also a part of [this PR](https://github.com/concept-to-clinic/concept-to-clinic/pull/236) that also shows currently running case, allows user to import and start new case after preview image, as it was described in [documentation](http://concept-to-clinic.readthedocs.io/en/latest/design-doc.html#imagery-interface-frontend).

**Comment by vessemer** [comment]
peek 2017-11-20 06-51

**Comment by lamby** [comment]
FYI I've started a discussion here <https://github.com/concept-to-clinic/concept-to-clinic/pull/236#issuecomment-345875065> regarding the potential feature conflict..

**Comment by vessemer** [comment]
peek 2017-11-23 02-38

File: interface/backend/api/views.py [hunk 2]
```diff
@@ -227,3 +227,42 @@ def nodule_update(request, nodule_id):
...
+    except Exception as e:
+        return Response({'response': "An error occurred: {}".format(e)}, 500)
+
+    image_series, created = ImageSeries.get_or_create(uri)
+    serialized = serializers.ImageSeriesSerializer(image_series, context={'request': request})
+    return Response(serialized.data)
+
+
+@api_view(['POST'])
+def case_available(request):
```
Comment (Serhiy-Shekhovtsov) [inline interface/backend/api/views.py]: There is no need to create a new method. `cases/` will do just fine.

File: interface/frontend/src/components/open-image/ImageSeries.vue [hunk 3]
```diff
@@ -8,21 +8,47 @@
...
+                <li v-for="el in cases" :key="el.created">
+                  <a href="#" @click="selectCase(el)">{{ el.created }}</a>
+                  <span v-if="el == selectedCase">&larr;</span>
                 </li>
               </ul>
+              <button @click="createCase(selected)"
+                      class="btn btn-primary float-left case-btn">
+                Create new case
+              </button>
+              <a href="#/report-and-export" class="btn btn-primary float-left case-btn">Start case</a>
```
Comment (Serhiy-Shekhovtsov) [inline interface/frontend/src/components/open-image/ImageSeries.vue]: As far as I understood from [this comment](https://github.com/concept-to-clinic/concept-to-clinic/pull/236#issuecomment-346512765) we should have multiple cases running same time. So there is no need to have **Start case** button. At least until there is no any role for it :)

File: interface/frontend/src/components/open-image/ImageSeries.vue [hunk 4]
```diff
@@ -8,21 +8,47 @@
...
             </template>
             <template v-else>
               <p class="card-text">No images imported.</p>
             </template>
             <button class="btn btn-warning float-right"
```
Comment (Serhiy-Shekhovtsov) [inline interface/frontend/src/components/open-image/ImageSeries.vue]: Please, remove this button(Show\Hide import) and make the tree view always visible, there is nothing else to show on this page. The only role of this page is to select a file and start the case, so there is no reason to make it require user to do extra clicks.

File: interface/frontend/src/components/open-image/ImageSeries.vue [hunk 5]
```diff
@@ -123,10 +136,28 @@
             // TODO: handle error
           })
       },
+      async fetchExistedCases (series) {
```
Comment (Serhiy-Shekhovtsov) [inline interface/frontend/src/components/open-image/ImageSeries.vue]: `fetchExistedCases` -> `fetchExistingCases`

File: interface/frontend/src/components/open-image/ImageSeries.vue [hunk 6]
```diff
@@ -141,6 +172,9 @@
 </script>
 
 <style lang="scss" scoped>
+  .case-btn {
+    margin-left: 1%
```
Comment (Serhiy-Shekhovtsov) [inline interface/frontend/src/components/open-image/ImageSeries.vue]: Bootstrap has `ml-1`, `ml-2` etc. for margins. Check [this](https://v4-alpha.getbootstrap.com/utilities/spacing/).

File: interface/frontend/src/components/open-image/ImageSeries.vue [hunk 7]
```diff
@@ -141,6 +172,9 @@
 </script>
 
 <style lang="scss" scoped>
+  .case-btn {
+    margin-left: 1% 
+  }
   .left {
     float: left;
```
Comment (Serhiy-Shekhovtsov) [inline interface/frontend/src/components/open-image/ImageSeries.vue]: You can use Bootstrap class `.pull-left` for this.

**Comment by Serhiy-Shekhovtsov** [comment]
One more thing, since you have fixed the DICOM viewer - user can scroll through slices now. So, clicking on last folder should actually show the image. No need to show list of files. You can merge my [code here](https://github.com/concept-to-clinic/concept-to-clinic/pull/236/files#diff-c93691d2fd82606ab2bb19978545678fL2).

**Comment by lamby** [comment]
Thanks for reviewing @Serhiy-Shekhovtsov  :)

**Comment by lamby** [comment]
Hey @vessemer, I assume you got all these comments, etc.? :)

**Comment by lamby** [comment]
No rush of course but the build seems to be failing and, alas, we need to resolve these conflicts :)

**Comment by lamby** [comment]
Gentle ping on this? :)

**Comment by Serhiy-Shekhovtsov** [comment]
I can reopen and update my PR if it will help us to proceed on this.

**Comment by Serhiy-Shekhovtsov** [comment]
@lamby, @reubano, I can reopen and update my PR if it will help us to proceed on this.

**Comment by reubano** [comment]
@Serhiy-Shekhovtsov no worries. tests are now passing and conflicts have been fixed.

**Comment by reubano** [comment]
Well, for whatever reason I'm not able to duplicate your screencasts. Here's what I get.

screen shot 2017-11-30 at 23 10 44
screen shot 2017-11-30 at 23 14 22

**Comment by Serhiy-Shekhovtsov** [comment]
Great job @vessemer! Few small comments from my side: 
- there are two buttons on the page: **Create case** and **Start case** and one of them does nothing but confuses the user :)
- also I think we should not show the list of files under last folder. Because once we switch to real data - we'll see **hundreds of files** there. It will make it complicated for user to browse the tree and this list doesn't do any help, cause the DICOM viewer should already support scrolling through slices. Again, you can copy\merge my [code here](https://github.com/concept-to-clinic/concept-to-clinic/pull/236/files#diff-c93691d2fd82606ab2bb19978545678fL2).

**Comment by reubano** [comment]
Turns out I forgot the `--hard` option in my `git reset` command. I'm getting the changed code now. Is there a data import step I need to run first?

**Comment by lamby** [comment]
@vessemer Ping on this? :)

**Comment by reubano** [comment]
Anyone else seeing an `[HPM] Error`? I'm still getting it. Even on `master`.

**Comment by reubano** [comment]
Re-ran everything and its looking good!

**Comment by reubano** [comment]
@Serhiy-Shekhovtsov great suggestions above. I've just fixed the conflicts so this PR should be merged soon. Mind submitting issues and accompanying PRs of your ideas?

**Comment by reubano** [comment]
Great job @vessemer! Once the tests complete (and hopefully pass!) do you mind rebasing everything to 1 commit? Thanks!

**Comment by Serhiy-Shekhovtsov** [comment]
_We should not show the list of files under last folder. Because once we switch to real data - we'll see **hundreds of files** there. It will make it complicated for user to browse the tree and this list doesn't do any help, cause the DICOM viewer should already support scrolling through slices._

This is still not fixed, right?

**Comment by reubano** [comment]
@Serhiy-Shekhovtsov correct.

**Comment by reubano** [comment]
There's a `candidate_dismiss is not defined` error in `interface/backend/api/urls.py`.

**Comment by isms** [comment]
FYI, @pjbull and I are in the midst of a cleanup and refactor right now that will take these changes into account - we can mark this is semi-accepted (thanks @vessemer and @Serhiy-Shekhovtsov for all the footwork). We'll credit appropriately.

[cc @reubano]

**Comment by lamby** [comment]
@isms Don't forget to re-ping here once done so we can resolve the conflicts :)

**Comment by isms** [comment]
@lamby All set from my perspective - we can close this and acknowledge contributors with full points. (This PR was really helpful in framing how we wanted things to work here, it just ended up being too far from the major refactored version to merge as-is. It otherwise would have been merged :+1:)

**Comment by lamby** [comment]
@isms  Done! Many thanks all o/

**Pull request closed without merging**